File: yukiko/c++/16_add_power_of_number.cpp
#include <iostream>
#include <vector>

using namespace std;

#define REP(i,first,last) for (int i=first;i<last;i++)

const int devider = 1000003;
int x, N;
vector<int> dp(devider-1, -1);

int my_pow(long long x, int idx){
  int res = 1; 
  if (idx > 0) {
    res = my_pow(x*x % devider, idx/2);
    if (idx & 1) res = res * x % devider;
  }
  return res % devider;
}

int main(){
  cin >> x;
  cin >> N;
  x %= devider;

  int val;
  int amari;
  long long total = 0;

  REP(i,0,N) {
    cin >> val;
    amari = val % (devider - 1);
    if (dp[amari] == -1) {
      dp[amari] = my_pow(x, amari);
    }
    total += dp[amari];
    total %= devider;
  }

  cout << (total % devider) << endl;
}

File: yukiko/ruby/16_add_power_of_number.rb
class Solve
  def initialize
    @devider = 1000003
    @x, _ = gets.chomp.split.map(&:to_i)
    @list = gets.chomp.split.map{ |i| i.to_i % (@devider - 1) }
    @dp = Array.new(@devider, -1)
    
    @x %= @devider
  end

  def run
    result = 0
    @list.each do |l|
      if @dp[l] == -1
        @dp[l] = my_pow(@x, l)
      end
      result += @dp[l]
    end

    p result % @devider
  end

  private

  def my_pow(n, idx)
    res = 1
    if idx > 0
      res = my_pow(n*n % @devider, idx/2)
      res = res * n % @devider if idx & 1 == 1
    end
    return res % @devider
  end
end

Solve.new.run
